src/test/java/net/openhft/chronicle/core/io/TracingReferenceCountedTest.java

```diff
@@ -0,0 +1,78 @@
+package net.openhft.chronicle.core.io;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static org.junit.Assert.*;
+
+public class TracingReferenceCountedTest extends MonitorReferenceCountedContractTest {
+
+    private AtomicInteger onReleaseCallCount;
+
+    @Before
+    public void setUp() {
+        onReleaseCallCount = new AtomicInteger(0);
+    }
+
+    @Override
+    protected TracingReferenceCounted createReferenceCounted() {
+        return new TracingReferenceCounted(onReleaseCallCount::incrementAndGet, "uniqueId", TracingReferenceCounted.class);
+    }
+
+    @Test
+    public void reserveWillThrowAndNotReserveWhenReferenceOwnerAttemptsToMakeASecondReservation() {
+        TracingReferenceCounted referenceCounted = createReferenceCounted();
+
+        ReferenceOwner a = ReferenceOwner.temporary("a");
+        referenceCounted.reserve(a);
+
+        assertEquals(2, referenceCounted.refCount());
+        assertThrows(IllegalStateException.class, () -> referenceCounted.reserve(a));
+        assertEquals(2, referenceCounted.refCount());
+    }
+
+    @Test
+    public void releaseWillFailWhenResourceOwnerHasNoReservation() {
+        TracingReferenceCounted referenceCounted = createReferenceCounted();
+
+        ReferenceOwner a = ReferenceOwner.temporary("a");
+        assertThrows(IllegalStateException.class, () -> referenceCounted.release(a));
+    }
+
+    /**
+     * This is actually a deviation from the contract, it should return false rather than throwing
+     * <p>
+     * Should we fix the contract or fix the behaviour?
+     */
+    @Test
```

Comment:
Tracing and Vanilla both throw `IllegalStateExceptions` when `reservedBy` is called with an owner that doesn't have a reservation. The interface says they should return false, we should change the implementations or the javadoc to be consistent. I've changed neither so far.

Reply:
The interface should show the contract that all implementations follow. The Vanilla can't detect duplicate reserves so it can't throw an ISE. The contract could say it may or may not throw an ISE depending on settings.

Reply:
The interface states that the two possible return values are `true`/`false` (as you'd expect given it returns a boolean). There is no mention of any exceptional cases. It accounts for the fact that some implementations may not know by stating

> Returns <code>true</code> if not sure.

(see below)
```
    /**
     * Best effort check the owner has reserved it. Returns <code>true</code> if not sure.
     *
     * @param owner to check
     * @return false if the owner definitely doesn't own it.
     */
```

I think we should never throw an exception, otherwise if someone wants to know if something is reserved by or not they need to implement
```
boolean actuallyReservedBy() {
    try {
        return bytes.reservedBy(owner);
    } catch (IllegalStateException e) {
        return false;
    }
}
```
which seems pretty clunky. (also assumes that exception means definitely not, as it seems to)

I checked where it's used (see [here](https://github.com/search?p=2&q=org%3AOpenHFT+reservedBy&type=Code), [here](https://github.com/search?q=org%3AChronicleEnterprise+reservedBy&type=Code) and [here](https://github.com/search?q=org%3AChronicleSoftware+reservedBy&type=Code)) and it's only used in assertions, if we return false instead of throwing an exception, we just get an `AssertionError` instead of and `IllegalStateException` when assertions are turned on.

What do you think?

Reply:
Another option might be to deprecate the current method and change it to `assertReservedBy(owner)`, which could throw an exception for the definitely not case and not return anything. (i.e. current behaviour without the redundant return value)

Reply:
I did the change to make them consistent with the current interface (i.e. not throwing exceptions). The tracing reference counted will still log a warning when false to distinguish between never-held and held-then-released. Running it through build-all-branch now.

Reply:
It's worth considering that the common case is that this level of tracing is off, and this is only useful when tracing is turned on. It's possible this method is no longer useful.

Reply:
I think if we just make the implementation match the interface as I've done, I'm happy to leave it there or it's becoming scope creep. There's also [a bunch of logic in TracingReferenceCounted](https://github.com/OpenHFT/Chronicle-Core/blob/b546716e3ef70139567886a3d6873c3167d7edff/src/main/java/net/openhft/chronicle/core/io/TracingReferenceCounted.java#L180) that looks like it could do with a tidy up and some more testing, but again it's scope creep for this PR. I'm happy where it's at for now and if we want to do more I suggest we create issues and more PRs later.